library/core/src/num/mod.rs

```diff
@@ -846,37 +870,32 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if radix <= 16 && src.len() <= mem::size_of::<T>() * 2 - (is_signed_ty && is_positive) as usize
+    {
+        // SAFETY: Consider the highest radix of 16:
+        // `u8::MAX` is `ff` which is 2 characters, `u16::MAX` out is `ffff` which is 4 characters
+        // We can be sure that any src len of 2 would fit in a u8 so we don't need
+        // to check for overflow.
+        unsafe {
+            let additive_op = if is_positive { T::unchecked_add } else { T::unchecked_sub };
+
+            for &c in digits {
+                result = result.unchecked_mul(radix);
+                let x = (c as char).to_digit(radix).ok_or_else(|| PIE { kind: InvalidDigit })?;
```

Comment:
Will using `?` here create a worse assembly? In a lot of cases match is still kept to provide better assembly, not sure if that's the case here.

Reply:
Seems like that's not the case any more: https://rust.godbolt.org/z/MPh4fGEfY
Unless I've missed something it's one xor instruction _less_ if we are not using `match`.

Reply:
Since we're not calling any expensive method, all those `ok_or_else` can be simplified to `ok_or` without affecting the resulting assembly.

Reply:
Correction: There is a difference for the checked branch, but not up here in the unchecked branch.